README.md
Helpy: A Modern Helpdesk Platform
====================================

Helpy is a modern help desk platform written in Ruby on Rails and released under the MIT license.  The goal of Helpy is to power your support email and ticketing, integrate seamlessly with your app, and run an amazing customer helpcenter.

[![Build Status](https://img.shields.io/travis/helpyio/helpy/master.svg)](https://travis-ci.org/helpyio/helpy) [![Code Climate](https://codeclimate.com/github/helpyio/helpy/badges/gpa.svg)](https://codeclimate.com/github/helpyio/helpy)

![](https://helpy.io/images/github-main-image.png)


Sponsor/Support Helpy
========

Helpy is licensed under the MIT license, and is an open-core project. This means that the core functionality is 100% open source and fully hackable or even re-sellable under the MIT license.  See the features comparison below to understand what is included.

Helpy is a large system and cannot exist purely as a hobby project. If you use it in a money generating capacity, it makes good sense to support the project financially or by becoming an official sponsor or supporter.

https://www.patreon.com/helpyio

Open Source Features
========

Helpy is an integrated support solution- combining and leveraging synergies between support ticketing, Knowledgebase and a public community.  Each feature is optional however, and can be easily disabled.

- **Multichannel ticketing:** Integrated with inbound email via Sendgrid, Mandrill, Mailgun, etc.
- **Knowledgebase:** Full text searchable and SEO optimized to help users answer questions before they contact you.
- **Mobile-friendly:** Support requests come at all times, and Helpy works on all devices out of the box so you can delight customers with prompt answers, from anywhere and at anytime!
- **Community Support Forums:** Customers and Agents can both answer questions in a publicly accessible forum, and vote both threads and replies up or down accordingly.
- **Embed Widget:** Helpy Includes a lightweight javascript widget that allows your users to contact you from just about anywhere.
- **Multi-lingual:** Helpy is fully multi-lingual and can provide support in multiple languages at the same time.  Currently the app includes translations for 19 languages and is easy to translate.
- **Themeable:** Customize the look and functionality of your Helpy without disturbing the underlying system that makes it all work. Helpy comes with two additional themes, and we hope to add more and get more from the community as time goes on.
- **Sends HTML email:** Responses to customers can include html, emojis and attachments.
- **Customizable:** Set colors to match your brand both on the helpcenter, and in the ticketing UI.
- **GDPR Compliant:** Comply with GDPR right to be forgotten requests by deleting users and their history, or by anonymizing them.

Pro Version
=========

We also offer a pro version with additional features designed to make your helpcenter even more awesome. This is available as either a turn-key SaaS or AWS/Azure marketplace product.  Both spin up in seconds. Proceeds go directly towards supporting the continued development of the project. Some of the things found in the pro version:

- **Triggers:** Insert events at any point in the ticket lifecycle. This includes an outbound JSON API.
- **Notifications:** Browser notifications when new tickets are received, you are assigned to a ticket, etc.
- **Real time UI:** When tickets arrive, they are automatically added to the UI
- **Custom Views:** Add additional Ticketing queues to highlight just the tickets you need to see
- **Advanced reporting:** A suite of additional reports on the performance of your ticketing and helpcenter knowledgebase
- **Advanced search:** Easily filter and find tickets or customers when you have thousands
- **Customizable Request Forms:** Easily Add questions to the ticket creation forms
- **AI Support Chatbot:** Create a chatbot for your website to answer up 90% of tier one questions autonomously


Getting Started:
=========

**Helpy Pro - 30 Second one-click install via AWS**

You can take a 14 day free trial of the Pro version of Helpy on AWS.  This is a "one click" marketplace image and spins up a dedicated VM within 30 seconds, running Ubuntu. [Launch AWS Marketplace Image](https://aws.amazon.com/marketplace/pp/B07WCNQGZ3?ref=_ptnr_web_github)

**Install Helpy via Docker**

Docker is the recommended way to quickly test or run Helpy in production.

1) Install [Docker](https://get.docker.com/) and docker-compose
2) Create config file from template `cp docker/.env.sample docker/.env` and edit `docker/.env` to match your needs
3) Edit `docker/Caddyfile` to include your URL or turn on SSL
4) Build Helpy from local git checkout `docker-compose build`
5) Run `docker-compose up -d` to start all of the services

**Install Helpy on your Local System**

Although not required, installing locally is highly recommended and will make it easier for you to customize things like settings, colors and logos to match your site identity.  To begin, clone Helpy from the official repo to your local system:

`git clone https://github.com/helpyio/helpy.git`

**Configure Basic Settings**

There is a settings option in the admin panel to set up things like i18n, system names, colors, the embeddable widget, etc.  There is a full guide to getting set up at: [Configuring Your Helpy Settings](http://support.helpy.io/en/knowledgebase/11-Installing-Helpy/docs/22-Configuring-your-Helpy-Settings)

**Support Multiple Languages (optional)**

Helpy includes support for Multilingual help sites, and multi-language knowledgebase articles.  This page explains how to enable Helpy's international capabilities and provides an overview of what functionality this adds to Helpy: [How To Set Up A Multilingual Helpy Support Knowledgebase](http://support.helpy.io/en/knowledgebase/12-Using-Helpy/docs/9-How-to-set-up-a-multilingual-Helpy-support-knowledgebase)

**Set up your Helpy to send and receive email (optional)**

Helpy has the ability to receive email at your support email addresses and import the messages as tickets in the web interface.  Whenever you or the user replies to the email thread, Helpy captures the discussion and sends out messages accordingly. Follow the tutorial on [Setting Up Your Helpy Installation To Send And Receive Email](http://support.helpy.io/en/knowledgebase/11-Installing-Helpy/docs/14-Setting-up-your-Helpy-installation-to-send-and-receive-email) to set this up.

**Configure oAuth (optional)**

Helpy supports Omniauth login capabilities.  This means you can allow your support users to sign in with a single click via any Omniauth provider- ie. Facebook, Twitter, Gmail, or many others. Read [Setting Up Oauth For Your Helpy](https://github.com/helpyio/helpy/wiki/SSO--Configuring-OAUTH) to see how.

Live Demo
=========

There is also a live demo with fake data available at [http://demo.helpy.io](http://demo.helpy.io)
Admin User: `admin@test.com` and password: `12345678`

Installation
============

Helpy was designed to run on modern cloud providers, although it should work on
any linux based system.  There is a full guide to installing Helpy in the wiki: https://github.com/helpyio/helpy/wiki

Requirements are:

- Ruby 2.4+
- Rails 4.2.x
- Postgres
- A server like Unicorn, Puma or Passenger

Helpy leverages two external services to help out:

- an email provider like Sendgrid
- Google Analytics for stats (optional)


Contributing
============

Welcome, and thanks for contributing to Helpy.  Together we are building the best customer support platform in the world.  Here are some of the ways you can contribute:

- Report or fix Bugs
- Refactoring
- Improve test coverage-  As with any large and growing codebase, test coverage is not always as good as it could be.  Help improving test coverage is always welcome and will help you learn how Helpy works.  We use Minitest exclusively.
- Translate the project- The community has already translated Helpy into 18 languages, but there are many more waiting.  We need help getting Helpy translated into as many locales as possible! [Please see the guide to translation](https://github.com/helpyio/helpy/wiki/How-to-translate-Helpy-into-your-language) for more details.
- Build new features.  There is a backlog of new features that we’d like to see built.  Check out our [roadmap](https://trello.com/b/NuiWsdmK/helpy) for more insight on this, and if you would like to take one on, please get in touch with us to make sure someone is not already working on it.

**General Guidelines:**

- Join us on Slack.  Let me know you wish to contribute. [![Slack Status](https://helpyioslackin.herokuapp.com/badge.svg)](https://helpyioslackin.herokuapp.com)
- Make your PRs granular.  They should only include one piece of functionality per PR.
- Check the roadmap: [Trello](https://trello.com/b/NuiWsdmK/helpy) If you want to build a feature, contact us to make sure no one else is already working on it
- You must provide passing test coverage.  We use minitest, see http://www.rubypigeon.com/posts/minitest-cheat-sheet/?utm_source=rubyweekly&utm_medium=email
- You also must expose functionality to the API.  We use Grape.  API methods should be tested as well.
- If your feature/bug fix/enhancement adds or changes text in the project, please create i18n strings in `en.yml` and any other locales you can.
- We are hugely concerned with user experience, and a nice UI.  Oftentimes that means we may take what you have contributed and “dress it up” or ask you to do the same.

Security Issues

CHANGELOG.md
## VERSION 2.8.0

Version 2.8 contains important security updates and the following new featuers:

- NEW It is now possible create users one at a time, with or without inviting them.
- NEW When creating tickets by API, you can now specify the CC and BCC for the new ticket.
- NEW A new API for viewing and editing taggings has been added (thanks @schmidt)
- NEW A honeypot (anti spam) feature was added to the new ticket form (thanks @schmidt)
- CHANGE Assigning an agent no longer automatically marks the ticket as "pending".
- CHANGE The whitelist of allowable file attachments has been significantly changed and made more flexible (see upgrade notes below).
- FIX A bug was fixed that displays the proper error message when a non supported file attachment is made in the web UI
- FIX #1576 The right menu no longer gets stuck open when turbolinks is enabled
- FIX #1581 Radio buttons are now properly aligned in the mobile responsive view of the Singular theme

IMPORTANT UPGRADE NOTES:

This release of Helpy includes a change to the way file attachments are handled.  Previous
versions used a default "whitelist" of file types where were allowed and was generally limited to
the most common image and doc foramts.

This version adds configuration settings which can be modified by API that allow you to set your own
whitelist of filetypes which should be allowed, or a blacklist of filetypes that should be rejected.  In
addition, the default whitelist has been removed, and Helpy now ships with a short blacklist of files that could be 
considered "risky."  Files with these extensions are blacklisted by default:

```
ade, adp, apk, appx, appxbundle, bat, cab, chm, cmd, com, cpl, dll, dmg, exe, hta, ins, isp, iso, jar, js, jse, lib, 
lnk, mde, msc, msi, msix, msixbundle, msp, mst, nsh, pif, ps1, scr, sct, .shb, sys, vb, vbe, vbs, vxd, wsc, wsf, wsh
```

NOTE:  Do not provide values for both blacklist and whitelist, as they will conflict and it will be impossible for
customers to attach files.


## VERSION 2.7.0

The 2.7 release of Helpy is here, with several great new features to help you better provide great customer support.

New Features:

- New nested categories feature lets you create sub categories with a drag and drop UI to group and administer them.
- New API setting for default reply type
- Site URL is now validated
- New Anti-spam setting to filter any ticket from matching email addresses to spam
- Address a ticket is sent to is now captured

Bug Fixes:

- Webhooks preference did not always save
- Customers list view does not always load

See UPDATING.md for details on how to update.

## VERSION 2.6.0

This release fixes several edge case bugs and adds a couple new API features and an all new Ukrainian language translation.  Specifically:

- New Ukrainian locale #1378 @Serg2294 

Bug Fixes:

- Translated versions are now properly included in the search index #1375 
- The CC address is now persisted if you change the assignment of a ticket while editing the reply #1387
- Agents can now create an unassigned ticket #1317
- A rare bug that docs unclickable on mobile has been resolved #1371 

API:

- A new endpoint has been added to get specifics of the currently logged in user #1242 @trevormerritt 
- Categories can now be filtered by visibility (public/private/internal) #1269
- Forum topics now respect a limit param #1357 @elguitar 
- The User entity is now included with topics and posts #1318 

Misc:

- Can now specify a theme environment var when running test suite #1253 @azul 


## VERSION 2.5.0

This release adds important new functionality for managing incoming spam email tickets (particularly from parse webhook services such as Sendgrid or Mandrill). 
Two new settings have been added that use the spam assassin score provided by these providers to either block tickets outright, or filter them to the spam
folder. In addition, dependencies have been upgraded.

- Expose trash and Spam tickets navigation in the left ticketing nav
- Adds two settings to adjust how Helpy filters spam (when using Sendgrid, Mandrill, etc.)
- Stores the spam score of incoming emails in the Topic table
- Adds the ability to do bulk actions on *all* tickets matching a search
- Adds the option to "empty the trash", permanently deleting all messages in the trash

See UPDATING.md for details on how to update.

## VERSION 2.4.3

This release fixes a bug in the way autosaving was handled. 

- Autosave functionality improved
- Added a button to clear autosaved draft if present
- Removed autosave from KB editor
- Depenedencies updated

## VERSION 2.4.2

This release updates dependecies and fixes an issue with the docker container which prevented it from booting.

- Dependencies updated
- Critical security update for Nokogiri

db/schema.rb
# encoding: UTF-8
# This file is auto-generated from the current state of the database. Instead
# of editing this file, please use the migrations feature of Active Record to
# incrementally modify your database, and then regenerate this schema definition.
#
# Note that this schema.rb definition is the authoritative source for your
# database schema. If you need to create the application database on another
# system, you should be using db:schema:load, not running all the migrations
# from scratch. The latter is a flawed and unsustainable approach (the more migrations
# you'll amass, the slower it'll run and the greater likelihood for issues).
#
# It's strongly recommended that you check this file into your version control system.

ActiveRecord::Schema.define(version: 20191005134018) do

  # These are extensions that must be enabled in order to support this database
  enable_extension "plpgsql"

  create_table "api_keys", force: :cascade do |t|
    t.string   "access_token"
    t.integer  "user_id"
    t.string   "name"
    t.datetime "date_expired"
    t.datetime "created_at",   null: false
    t.datetime "updated_at",   null: false
  end

  add_index "api_keys", ["access_token"], name: "index_api_keys_on_access_token", unique: true, using: :btree
  add_index "api_keys", ["user_id"], name: "index_api_keys_on_user_id", using: :btree

  create_table "attachinary_files", force: :cascade do |t|
    t.integer  "attachinariable_id"
    t.string   "attachinariable_type"
    t.string   "scope"
    t.string   "public_id"
    t.string   "version"
    t.integer  "width"
    t.integer  "height"
    t.string   "format"
    t.string   "resource_type"
    t.datetime "created_at"
    t.datetime "updated_at"
  end

  add_index "attachinary_files", ["attachinariable_type", "attachinariable_id", "scope"], name: "by_scoped_parent", using: :btree

  create_table "audits", force: :cascade do |t|
    t.integer  "auditable_id"
    t.string   "auditable_type"
    t.integer  "associated_id"
    t.string   "associated_type"
    t.integer  "user_id"
    t.string   "user_type"
    t.string   "username"
    t.string   "action"
    t.text     "audited_changes"
    t.integer  "version",         default: 0
    t.string   "comment"
    t.string   "remote_address"
    t.string   "request_uuid"
    t.datetime "created_at"
  end

  add_index "audits", ["associated_id", "associated_type"], name: "associated_index", using: :btree
  add_index "audits", ["auditable_id", "auditable_type"], name: "auditable_index", using: :btree
  add_index "audits", ["created_at"], name: "index_audits_on_created_at", using: :btree
  add_index "audits", ["request_uuid"], name: "index_audits_on_request_uuid", using: :btree
  add_index "audits", ["user_id", "user_type"], name: "user_index", using: :btree

  create_table "backups", force: :cascade do |t|
    t.integer  "user_id"
    t.text     "csv"
    t.string   "model"
    t.string   "csv_name"
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
  end

  add_index "backups", ["user_id"], name: "index_backups_on_user_id", using: :btree

  create_table "boxes", force: :cascade do |t|
    t.string   "label"
    t.string   "name"
    t.text     "description"
    t.integer  "rank"
    t.text     "query"
    t.boolean  "default"
    t.datetime "created_at",  null: false
    t.datetime "updated_at",  null: false
  end

  create_table "categories", force: :cascade do |t|
    t.string   "name"
    t.string   "icon"
    t.string   "keywords"
    t.string   "title_tag"
    t.string   "meta_description"
    t.integer  "rank"
    t.boolean  "front_page",       default: false
    t.boolean  "active",           default: true
    t.string   "permalink"
    t.string   "section"
    t.datetime "created_at",                       null: false
    t.datetime "updated_at",                       null: false
    t.string   "visibility",       default: "all"
    t.string   "ancestry"
  end

  add_index "categories", ["ancestry"], name: "index_categories_on_ancestry", using: :btree

  create_table "category_translations", force: :cascade do |t|
    t.integer  "category_id",      null: false
    t.string   "locale",           null: false
    t.datetime "created_at",       null: false
    t.datetime "updated_at",       null: false
    t.string   "name"
    t.string   "keywords"
    t.string   "title_tag"
    t.string   "meta_description"
  end

  add_index "category_translations", ["category_id"], name: "index_category_translations_on_category_id", using: :btree
  add_index "category_translations", ["locale"], name: "index_category_translations_on_locale", using: :btree

  create_table "doc_translations", force: :cascade do |t|
    t.integer  "doc_id",           null: false
    t.string   "locale",           null: false
    t.datetime "created_at",       null: false
    t.datetime "updated_at",       null: false
    t.string   "title"
    t.text     "body"
    t.string   "keywords"
    t.string   "title_tag"
    t.string   "meta_description"
  end

  add_index "doc_translations", ["doc_id"], name: "index_doc_translations_on_doc_id", using: :btree
  add_index "doc_translations", ["locale"], name: "index_doc_translations_on_locale", using: :btree

  create_table "doc_views", force: :cascade do |t|
    t.string   "collector_action", default: "view"
    t.integer  "user_id"
    t.string   "session_id"
    t.integer  "doc_id"
    t.string   "doc_title"
    t.integer  "category_id"
    t.string   "doc_category"
    t.string   "search_used"
    t.text     "referrer"
    t.datetime "created_at",                        null: false
    t.datetime "updated_at",                        null: false
    t.integer  "results_found",    default: 0
  end

  add_index "doc_views", ["doc_id"], name: "index_doc_views_on_doc_id", using: :btree

  create_table "docs", force: :cascade do |t|
    t.string   "title"
    t.text     "body"
    t.string   "keywords"
    t.string   "title_tag"
    t.string   "meta_description"
    t.integer  "category_id"
    t.integer  "user_id"
    t.boolean  "active",           default: true
    t.integer  "rank"
    t.string   "permalink"
    t.integer  "version"
    t.boolean  "front_page",       default: false
    t.boolean  "cheatsheet",       default: false
    t.integer  "points",           default: 0
    t.datetime "created_at",                       null: false
    t.datetime "updated_at",                       null: false
    t.integer  "topics_count",     default: 0
    t.boolean  "allow_comments",   default: true
    t.string   "attachments",      default: [],                 array: true
    t.integer  "doc_views_count",  default: 0
  end

  create_table "flags", force: :cascade do |t|
    t.integer  "post_id"
    t.integer  "generated_topic_id"
    t.text     "reason"
    t.datetime "created_at",         null: false
    t.datetime "updated_at",         null: false
  end

  create_table "followers", force: :cascade do |t|
    t.integer  "user_id"
    t.integer  "topic_id"
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
  end

  add_index "followers", ["topic_id"], name: "index_followers_on_topic_id", using: :btree
  add_index "followers", ["user_id"], name: "index_followers_on_user_id", using: :btree

  create_table "forums", force: :cascade do |t|
    t.string   "name"
    t.text     "description"
    t.integer  "topics_count",       default: 0,       null: false
    t.datetime "last_post_date"
    t.integer  "last_post_id"
    t.boolean  "private",            default: false
    t.datetime "created_at",                           null: false
    t.datetime "updated_at",                           null: false
    t.boolean  "allow_topic_voting", default: false
    t.boolean  "allow_post_voting",  default: false
    t.string   "layout",             default: "table"
  end

  create_table "images", force: :cascade do |t|
    t.string   "key"
    t.string   "name"
    t.string   "extension"
    t.text     "file"
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
  end

  create_table "imports", force: :cascade do |t|
    t.string   "status"
    t.string   "notes"
    t.string   "model"
    t.datetime "started_at"
    t.datetime "completed_at"
    t.integer  "submited_record_count"
    t.text     "imported_ids"
    t.text     "error_log"
    t.datetime "created_at",            null: false
    t.datetime "updated_at",            null: false
  end

  create_table "key_values", force: :cascade do |t|
    t.string   "key"
    t.text     "value"
    t.integer  "kvable_id"
    t.string   "kvable_type"
    t.datetime "created_at",  null: false
    t.datetime "updated_at",  null: false
  end

  add_index "key_values", ["kvable_id", "kvable_type"], name: "kvable_index", using: :btree
  add_index "key_values", ["value"], name: "kv_value_index", using: :btree

  create_table "notifications", force: :cascade do |t|
    t.integer  "recipient_id"
    t.integer  "actor_id"
    t.datetime "read_at"
    t.boolean  "send_email",      default: false
    t.datetime "sent_at"
    t.string   "action"
    t.string   "title"
    t.text     "message"
    t.integer  "notifiable_id"
    t.string   "notifiable_type"
    t.datetime "created_at",                      null: false
    t.datetime "updated_at",                      null: false
    t.datetime "alerted_at"
  end

  create_table "pg_search_documents", force: :cascade do |t|
    t.text     "content"
    t.integer  "searchable_id"
    t.string   "searchable_type"
    t.datetime "created_at",      null: false
    t.datetime "updated_at",      null: false
  end

  create_table "posts", force: :cascade do |t|
    t.integer  "topic_id"
    t.integer  "user_id"
    t.text     "body"
    t.string   "kind"
    t.boolean  "active",           default: true
    t.datetime "created_at",                      null: false
    t.datetime "updated_at",                      null: false
    t.integer  "points",           default: 0
    t.string   "attachments",      default: [],                array: true
    t.string   "cc"
    t.string   "bcc"
    t.text     "raw_email"
    t.string   "email_to_address", default: ""
  end

  create_table "ratings", force: :cascade do |t|
    t.integer  "topic_id"
    t.integer  "score"
    t.text     "comments"
    t.integer  "user_id"
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
  end

  add_index "ratings", ["topic_id"], name: "index_ratings_on_topic_id", using: :btree
  add_index "ratings", ["user_id"], name: "index_ratings_on_user_id", using: :btree

  create_table "searches", force: :cascade do |t|
    t.string   "name"
    t.text     "body"
    t.string   "search_type"
    t.integer  "search_id"
    t.datetime "last_updated_at"
    t.datetime "created_at",      null: false
    t.datetime "updated_at",      null: false
  end

  create_table "settings", force: :cascade do |t|
    t.string   "var",                   null: false
    t.text     "value"
    t.integer  "thing_id"
    t.string   "thing_type", limit: 30
    t.datetime "created_at"
    t.datetime "updated_at"
  end

  add_index "settings", ["thing_type", "thing_id", "var"], name: "index_settings_on_thing_type_and_thing_id_and_var", unique: true, using: :btree

  create_table "sla_schedules", force: :cascade do |t|
    t.integer  "topic_id"
    t.integer  "sla_id"
    t.integer  "post_id"
    t.datetime "wait_until"
    t.datetime "job_run"
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
  end

  create_table "slas", force: :cascade do |t|
    t.string   "name"
    t.text     "description"
    t.string   "event"
    t.string   "priority"
    t.integer  "time"
    t.integer  "time_units",                 default: 60
    t.integer  "minutes"
    t.string   "tags"
    t.string   "group"
    t.integer  "assigned_user_id"
    t.text     "note"
    t.string   "notify_users",                                            array: true
    t.boolean  "active",                     default: true
    t.datetime "created_at",                                 null: false
    t.datetime "updated_at",                                 null: false
    t.string   "selected_groups",                                         array: true
    t.string   "selected_ticket_priorities",                              array: true
    t.string   "selected_user_priorities",                                array: true
    t.boolean  "set_hours",                  default: false
    t.boolean  "monday_active",              default: false
    t.integer  "monday_start",               default: 9
    t.integer  "monday_end",                 default: 18
    t.boolean  "tuesday_active",             default: false
    t.integer  "tuesday_start",              default: 9
    t.integer  "tuesday_end",                default: 18
    t.boolean  "wednesday_active",           default: false
    t.integer  "wednesday_start",            default: 9
    t.integer  "wednesday_end",              default: 18
    t.boolean  "thursday_active",            default: false
    t.integer  "thursday_start",             default: 9
    t.integer  "thursday_end",               default: 18
    t.boolean  "friday_active",              default: false
    t.integer  "friday_start",               default: 9
    t.integer  "friday_end",                 default: 18
    t.boolean  "saturday_active",            default: false
    t.integer  "saturday_start",             default: 9
    t.integer  "saturday_end",               default: 18
    t.boolean  "sunday_active",              default: false
    t.integer  "sunday_start",               default: 9
    t.integer  "sunday_end",                 default: 18
    t.string   "time_zone",                  default: "UTC"
  end

  create_table "taggings", force: :cascade do |t|
    t.integer  "tag_id"
    t.integer  "taggable_id"
    t.string   "taggable_type"
    t.integer  "tagger_id"
    t.string   "tagger_type"
    t.string   "context",       limit: 128
    t.datetime "created_at"
  end

  add_index "taggings", ["tag_id", "taggable_id", "taggable_type", "context", "tagger_id", "tagger_type"], name: "taggings_idx", unique: true, using: :btree
  add_index "taggings", ["taggable_id", "taggable_type", "context"], name: "index_taggings_on_taggable_id_and_taggable_type_and_context", using: :btree

  create_table "tags", force: :cascade do |t|
    t.string  "name"
    t.integer "taggings_count",     default: 0
    t.boolean "show_on_helpcenter", default: false
    t.boolean "show_on_admin",      default: false
    t.boolean "show_on_dashboard",  default: false
    t.text    "description"
    t.string  "color"
    t.boolean "active",             default: true
    t.string  "email_address"
    t.string  "email_name"
  end

  add_index "tags", ["name"], name: "index_tags_on_name", unique: true, using: :btree

  create_table "topic_fields", force: :cascade do |t|
    t.string   "name"
    t.string   "label"
    t.string   "field_type"
    t.string   "select_options"
    t.boolean  "required"
    t.string   "class_names"
    t.integer  "rank"
    t.string   "group"
    t.string   "toggle_target"
    t.boolean  "display_on_helpcenter"
    t.boolean  "display_on_admin"
    t.datetime "created_at",            null: false
    t.datetime "updated_at",            null: false
    t.string   "valid_value"
  end

  create_table "topics", force: :cascade do |t|
    t.integer  "forum_id"
    t.integer  "user_id"
    t.string   "user_name"
    t.string   "name"
    t.integer  "posts_count",      default: 0,        null: false
    t.string   "waiting_on",       default: "admin",  null: false
    t.datetime "last_post_date"
    t.datetime "closed_date"
    t.integer  "last_post_id"
    t.string   "current_status",   default: "new",    null: false
    t.boolean  "private",          default: false
    t.integer  "assigned_user_id"
    t.boolean  "cheatsheet",       default: false
    t.integer  "points",           default: 0
    t.text     "post_cache"
    t.datetime "created_at",                          null: false
    t.datetime "updated_at",                          null: false
    t.string   "locale"
    t.integer  "doc_id",           default: 0
    t.string   "channel",          default: "email"
    t.string   "kind",             default: "ticket"
    t.integer  "priority",         default: 1
    t.string   "condition",        default: "green"
    t.string   "sentiment"
    t.datetime "last"
    t.decimal  "spam_score",       default: 0.0
    t.text     "spam_report",      default: ""
  end

  add_index "topics", ["kind"], name: "index_topics_on_kind", using: :btree
  add_index "topics", ["priority"], name: "index_topics_on_priority", using: :btree

  create_table "trigger_runs", force: :cascade do |t|
    t.integer  "topic_id"
    t.integer  "trigger_id"
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
  end

  create_table "triggers", force: :cascade do |t|
    t.string   "name"
    t.string   "mode",          default: "and"
    t.text     "actions"
    t.text     "conditions"
    t.string   "url"
    t.string   "event"
    t.string   "slack_channel"
    t.boolean  "active",        default: true
    t.integer  "rank"
    t.datetime "created_at",                    null: false
    t.datetime "updated_at",                    null: false
  end

  create_table "users", force: :cascade do |t|
    t.string   "login"
    t.string   "identity_url"
    t.string   "name"
    t.boolean  "admin",                  default: false
    t.text     "bio"
    t.text     "signature"
    t.string   "role",                   default: "user"
    t.string   "home_phone"
    t.string   "work_phone"
    t.string   "cell_phone"
    t.string   "company"
    t.string   "street"
    t.string   "city"
    t.string   "state"
    t.string   "zip"
    t.string   "title"
    t.string   "twitter"
    t.string   "linkedin"
    t.string   "thumbnail"
    t.string   "medium_image"
    t.string   "large_image"
    t.string   "language",               default: "en"
    t.integer  "assigned_ticket_count",  default: 0
    t.integer  "topics_count",           default: 0
    t.boolean  "active",                 default: true
    t.datetime "created_at",                                   null: false
    t.datetime "updated_at",                                   null: false
    t.string   "email",                  default: "",          null: false
    t.string   "encrypted_password",     default: "",          null: false
    t.string   "reset_password_token"
    t.datetime "reset_password_sent_at"
    t.datetime "remember_created_at"
    t.integer  "sign_in_count",          default: 0,           null: false
    t.datetime "current_sign_in_at"
    t.datetime "last_sign_in_at"
    t.inet     "current_sign_in_ip"
    t.inet     "last_sign_in_ip"
    t.string   "provider"
    t.string   "uid"
    t.string   "invitation_token"
    t.datetime "invitation_created_at"
    t.datetime "invitation_sent_at"
    t.datetime "invitation_accepted_at"
    t.integer  "invitation_limit"
    t.integer  "invited_by_id"
    t.string   "invited_by_type"
    t.integer  "invitations_count",      default: 0
    t.text     "invitation_message"
    t.string   "time_zone",              default: "UTC"
    t.string   "profile_image"
    t.boolean  "notify_on_private",      default: false
    t.boolean  "notify_on_public",       default: false
    t.boolean  "notify_on_reply",        default: false
    t.string   "account_number"
    t.string   "priority",               default: "normal"
    t.text     "notes"
    t.boolean  "notify_on_assignment",   default: true
    t.boolean  "notify_on_mention",      default: true
    t.string   "status",                 default: "available"
  end

  add_index "users", ["email"], name: "index_users_on_email", unique: true, using: :btree
  add_index "users", ["invitation_token"], name: "index_users_on_invitation_token", unique: true, using: :btree
  add_index "users", ["invitations_count"], name: "index_users_on_invitations_count", using: :btree
  add_index "users", ["invited_by_id"], name: "index_users_on_invited_by_id", using: :btree
  add_index "users", ["notify_on_private"], name: "index_users_on_notify_on_private", using: :btree
  add_index "users", ["notify_on_public"], name: "index_users_on_notify_on_public", using: :btree
  add_index "users", ["notify_on_reply"], name: "index_users_on_notify_on_reply", using: :btree
  add_index "users", ["priority"], name: "index_users_on_priority", using: :btree
  add_index "users", ["reset_password_token"], name: "index_users_on_reset_password_token", unique: true, using: :btree

  create_table "versions", force: :cascade do |t|
    t.string   "item_type",  null: false
    t.integer  "item_id",    null: false
    t.string   "event",      null: false
    t.string   "whodunnit"
    t.text     "object"
    t.datetime "created_at"
    t.string   "locale"
  end

  add_index "versions", ["item_type", "item_id"], name: "index_versions_on_item_type_and_item_id", using: :btree

  create_table "violations", force: :cascade do |t|
    t.integer  "topic_id"
    t.integer  "sla_id"
    t.boolean  "active"
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
  end

  create_table "votes", force: :cascade do |t|
    t.integer  "points",        default: 1
    t.string   "voteable_type"
    t.integer  "voteable_id"
    t.integer  "user_id",       default: 0
    t.datetime "created_at",                null: false
    t.datetime "updated_at",                null: false
  end

end
